File: node.js/sampleCode.js
const uuid = require('uuid');
const axios = require('axios').default;


/*
    Source env.rc to export some sensitive data to Environment Variables.
    Get these Variable here.
*/
const _API_STAGE = process.env.API_STAGE;
const _API_VERSION = process.env.API_VERSION;
const _CLIENT_ID = process.env.CLIENT_ID;
const _CLIENT_SECRET = process.env.CLIENT_SECRET;

/*
    Check the status of the API without any authentication headers.
*/
var checkAPIstatus = function () {
    return new Promise((resolve, reject) => {    
        const url = `https://${_API_STAGE}/base/${_API_VERSION}/ping`;        
        axios
            .get(url)
            .then(response => {
                resolve(response.data);
            })
            .catch(error => {
                reject(error);
            });
    });
}

/*
    Get chunk of marketplaceEvents.
    - limit: indicate the amount of data retrieves. recommended value is 50
    return value is an array of new added marketplaceEvents with maximum size of "limit".
*/
var getMarketplaceEvents = function (token, url, limit) {
    console.log(`Invoke API: ${url}?limit=${limit}`);
    return new Promise((resolve, reject) => {
        config = {
            params: {
                "limit": limit
            },
            headers: {
                "X-Request-ID": uuid.v4(),
                "Authorization": token
            }
        };
        axios
            .get(url, config)
            .then(response => {
                resolve(response);
            })
            .catch(error => {
                reject(error);
            });
    });
}


/*
    Get all of marketplaceEvents.
    This is a recursive function and call itself untill get all marketplaceEvents chunk by chunk.
    - limit: indicate the amount of data retrieves. recommended value is 50    
    - marketplaceEvents: An array that we fill it in each recursive function call.    
    return value is an array of all new added marketplaceEvents.
*/
var getMarketplaceEventsRecursive = function (token, url, limit, marketplaceEvents) {
    return new Promise((resolve, reject) => {
        getMarketplaceEvents(token, url, limit)
            .then(response => {
                response.data.marketplaceEvents.forEach(element => {
                    marketplaceEvents.push(element);
                });
                if(response.data.hasMore) {
                    resolve(getMarketplaceEventsRecursive(token, url, limit, marketplaceEvents));
                }else{
                    resolve(marketplaceEvents);
                }
            })
            .catch(error => {
                reject(error);
            });
    });
}


/*
    Get order by it's Id.    
    - orderId
    return value is an Order object.
*/
var getOrder = function (token, orderId) {
    const url = `https://${_API_STAGE}/lead/${_API_VERSION}/orders/${orderId}`;
    console.log(`Invoke API: ${url}`);
    return new Promise((resolve, reject) => {
        config = {
            headers: {
                "X-Request-ID": uuid.v4(),
                "Authorization": token
            }
        };
        axios
            .get(url, config)
            .then(response => {
                resolve(response);
            })
            .catch(error => {
                reject(error);
            });
    });
}


/*
    Get License by it's Id.    
    - licenseId
    return value is an License object.
*/
var getLicense = function (token, licenseId) {
    const url = `https://${_API_STAGE}/lead/${_API_VERSION}/licenses/${licenseId}`;
    console.log(`Invoke API: ${url}`);
    return new Promise((resolve, reject) => {
        config = {
            headers: {
                "X-Request-ID": uuid.v4(),
                "Authorization": token
            }
        };
        axios
            .get(url, config)
            .then(response => {
                resolve(response);
            })
            .catch(error => {
                reject(error);
            });
    });
}

/*
    Dispatch request to get Order or License based on additionalProperties in marketplaceEvent.
    You have to add your custom code to provision service or license here.
    - marketplaceEvent
    return value is an License object or an Order object.
*/
var dispatch = function (marketplaceEvent) {
    return new Promise((resolve, reject) => {
        additionalProperties = marketplaceEvent["additionalProperties"]
        additionalProperties.forEach(additionalProperty => {
            if (additionalProperty["name"] == "orderId") {
                getOrder(token, additionalProperty["value"])
                    .then(response => {
                        // This is Order
                        // Add Your provisioning code here
                        resolve('Success!')
                    })
                    .catch(error => {
                        reject(error)
                    });
            } else if (additionalProperty["name"] == "licenseId"){
                getLicense(token, additionalProperty["value"])
                    .then(response => {
                        // This is License
                        // Add Your license management code here
                        resolve('Success!')
                    })
                    .catch(error => {
                        reject(error)
                    });
            }    
        });
    });
}

/*
    pass ClientId and ClientSecret to this method and get time-limited JWT Bearer token.
    Token will be expired in 30 minutes.
*/
var authenticate = function () {
    return new Promise((resolve, reject) => {
        const url = `https://${_API_STAGE}/auth/${_API_VERSION}/tokens`;
        config = {
            headers: {
                'X-Request-ID': uuid.v4(),
                'Authorization': 'Basic ' + Buffer.from(`${_CLIENT_ID}:${_CLIENT_SECRET}`, 'utf8').toString('base64')
            }
        };
        axios
            .post(url, null, config)
            .then(response => {
                token = `${response.data.tokenType} ${response.data.accessToken}`;
                resolve(token);
            })
            .catch(error => {
                reject(error);
            });
    });
}


/*
    Call checkAPIstatus to see the simple ping response form API server.
*/
/*checkAPIstatus()
    .then((ping) => {
        console.log(ping);
    })
    .catch(error => {
        console.log(error);
    });
*/



/*
    1. First call authenticate function to have a JWT tokne.
    2. Get new marketplaceEvents by call getMarketplaceEventsRecursive
    3. For each marketplaceEvent, call Dispatch function.
*/
authenticate(_API_STAGE, _API_VERSION, _CLIENT_ID, _CLIENT_SECRET)
    .then((token) => {
        getMarketplaceEventsRecursive(token, `https://${_API_STAGE}/lead/${_API_VERSION}/marketplaceEvents`, 40, marketplaceEvents=[])
            .then((marketplaceEvents) => {
                marketplaceEvents.forEach(marketplaceEvent => {
                    dispatch(marketplaceEvent)
                        .then(response => {
                            console.log(response);
                        })
                        .catch(error => {
                            reject(error);
                        });
                });
            })
            .catch((error) => {
                console.log(error);
            });
    })
    .catch(error => {
        console.log(error);
    });

File: python/sampleCode.py
import os
import json
import requests
import uuid
import base64


"""
    Source env.rc to export some sensitive data to Environment Variables.
    Get these Variable here.
"""
_API_STAGE = os.getenv('API_STAGE', "")
_API_VERSION = os.getenv('API_VERSION', "")
_CLIENT_ID = os.getenv('CLIENT_ID', "")
_CLIENT_SECRET = os.getenv('CLIENT_SECRET', "")
INTERNAL_ERROR_CODE = 500

def base64wrapper(message):
    message_bytes = message.encode('ascii')
    base64_bytes = base64.b64encode(message_bytes)
    base64_message = base64_bytes.decode('ascii')
    return base64_message

"""
    Call checkAPIstatus to see the simple ping response form API server.
"""
def checkAPIStatus():
    try:
        endpoint = f'https://{_API_STAGE}/base/{_API_VERSION}/ping'
        response = requests.get(endpoint)
        return response.json(), response.status_code
    except Exception as e:
        print(f'{e}')
        return e, INTERNAL_ERROR_CODE

"""
    pass ClientId and ClientSecret to this method and get time-limited JWT Bearer token.
    Token will be expired in 30 minutes.
"""
def authenticate():
    try:
        endpoint = f'https://{_API_STAGE}/auth/{_API_VERSION}/tokens'
        headers = {
            "X-Request-ID": str(uuid.uuid4()),
            "Authorization": 'Basic ' + base64wrapper(f'{_CLIENT_ID}:{_CLIENT_SECRET}')
        }
        response = requests.post(endpoint, headers=headers, data={})
        if response.status_code == 201:
            return response.json().get("tokenType", "") + " " + response.json().get("accessToken", ""), response.status_code
        return response.json(), response.status_code
    except Exception as e:
        print(f'{e}')
        return e, INTERNAL_ERROR_CODE


"""
    Get chunk of marketplaceEvents.
    - limit: indicate the amount of data retrieves. recommended value is 50
    return value is an array of new added marketplaceEvents with maximum size of "limit".
"""
def getMarketplaceEvents(token, url, limit):
    marketplaceEvents = []
    hasMore = True
    try:
        while (hasMore):
            params = {
                "limit": limit
            }
            headers = {
                "X-Request-ID": str(uuid.uuid4()),
                "Authorization": token
            }
            print(f'Invoke API: {url}/?limit={limit}')
            response = requests.get(url, params=params, headers=headers)
            for marketplaceEvent in response.json().get("marketplaceEvents", []):
                marketplaceEvents.append(marketplaceEvent)
            hasMore = response.json().get("hasMore", False)
        return marketplaceEvents, response.status_code
    except Exception as e:
        print(f'{e}')
        return e, INTERNAL_ERROR_CODE


"""
    Get order by it's Id.    
    - orderId
    return value is an Order object.
"""
def getOrder(token, orderId):
    try:
        endpoint = f'https://{_API_STAGE}/lead/{_API_VERSION}/orders/{orderId}'
        headers = {
            "X-Request-ID": str(uuid.uuid4()),
            "Authorization": token
        }
        print(f'Invoke API: {endpoint}')
        response = requests.get(endpoint, headers=headers)
        orderObj = response.json()
        # Add Your code here
        return orderObj, response.status_code
    except Exception as e:
        print(f'{e}')
        return e, INTERNAL_ERROR_CODE


"""
License includes details about the buyer and the purchased product. Each license is identified by an identifier licenseId.
"""
def getLicense(token, licenseId):
    try:
        endpoint = f'https://{_API_STAGE}/lead/{_API_VERSION}/licenses/{licenseId}'
        headers = {
            "X-Request-ID": str(uuid.uuid4()),
            "Authorization": token
        }
        print(f'Invoke API: {endpoint}')
        response = requests.get(endpoint, headers=headers)
        licenseObj = response.json()
        # Add Your code here
        return licenseObj, response.status_code
    except Exception as e:
        print(f'{e}')
        return e, INTERNAL_ERROR_CODE

def dispatch(token, marketplaceEvent):
    additionalProperties = marketplaceEvent.get("additionalProperties", [])
    for additionalProperty in additionalProperties:
        if (additionalProperty.get("name", "") == "orderId"):
            return getOrder(token, additionalProperty.get("value", ""))
        elif (additionalProperty.get("name", "") == "licenseId"):
            return getLicense(token, additionalProperty.get("value", ""))

# checkAPIStatus
ping, status = checkAPIStatus()
if status != 200:
    print('ERROR: API call failed.')
    exit()
#print(ping)

# authenticate
token, status = authenticate()
if status != 201:
    print(f'ERROR: API call failed. {status}, {token}')
    exit()
#print(token)

# getMarketplaceEvents
marketplaceEvents, status = getMarketplaceEvents(
        token,
        f'https://{_API_STAGE}/lead/{_API_VERSION}/marketplaceEvents', 40)
if status != 200:
    print(f'ERROR: API call failed. {status}')
    exit()
#print(marketplaceEvents)
for marketplaceEvent in marketplaceEvents:
    dispatch(token, marketplaceEvent)
